### PR TITLE
Google Drive storage

### LINKED ISSUE #925: Google Drive Sync
We still have latent Google Drive sync code in DIM, but commented out. This issue is to either finish and release that feature, or remove it.

### PR DESCRIPTION
**Pull request by bhollis**
This is a work-in-progress PR to add Google Drive storage support to DIM. As of this PR, we have a (feature flagged) Google Drive storage adapter that works, but not a lot of nice UI or error handling around it. Part of this was refactoring the storage system into a series of generic adapters that are invoked in sequence, so we can plug in other methods. For example, I've added IndexedDB as well, as a nicer replacement for local storage. This will eventually fix #925.

While I was in here I also did a bunch of work to avoid touching the sync system if we don't have to (updating values to the same value, removing stuff that already doesn't exist, avoiding duplicate parallel "get"s, etc).

* [x] Figure out whether to have separate beta/prod storage or to use the same file.
* [x] Documentation
* [ ] Error handling (lots of it)
* [x] UI for logging in, logging out
* [x] UI for explaining storage
* [x] UI for import/export of data
* [x] Warnings when enabling Google Sync that you'll overwrite local data with whatever's in Drive
* [x] Reload data after authorizing with Google
* [ ] Save timestamps / versioning so we can resolve out of order updates, support offline saving that later syncs up (stretch)
* [ ] Defer loading the Google Drive adapter until after we have a platform set
* [x] Switch to a Bungie.net account-specific membership ID instead of Destiny account specific membership ID.
* [x] Remove local storage (with an upgrade)
* [x] Make the storage page actually nice

### CODE REVIEW

**Comment by delphiactual** [comment]
Just an idea:
settings should be saved per flavor
item-tags and loadouts should be universal between all flavors
and any other features we add should be universal just not loadable until the features are available in stable...

**Comment by bhollis** [comment]
I've added a rudimentary storage settings/diagnostics page - this could be its own route, as I have it now, or show in a modal (it's just a component after all). It's not pretty, but I think once it's cleaned up it'll be really valuable to have one place for users to go to figure out what they have saved, where it's saved, back up their data, etc.

**Comment by bhollis** [comment]
screen shot 2017-06-16 at 11 54 10 pm

File: src/scripts/storage/storage.component.js [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+  vm.syncService = SyncService;
+  vm.adapterStats = {};
+
+  function dataStats(data) {
+    const taggedItems = sum(Object.keys(data)
+                            .filter((k) => k.startsWith('dimItemInfo'))
+                            .map((k) => Object.keys(data[k]).length));
+
+    return {
+      Loadouts: data['loadouts-v3.0'].length,
```
Comment (delphiactual) [inline src/scripts/storage/storage.component.js]: Loadouts: data['loadouts-v3.0'] ? data['loadouts-v3.0'].length : 0;

File: src/scripts/storage/storage.component.js [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+  vm.importData = function() {
+    var reader = new FileReader();
+    reader.onload = function() {
+      // TODO: we're kinda trusting that this is the right data here, no validation!
+      SyncService.set(JSON.parse(reader.result), true);
+      $window.alert("Imported DIM data!");
+      $scope.$apply();
+    };
+    reader.readAsText(angular.element('#importFile')[0].files[0]);
```
Comment (delphiactual) [inline src/scripts/storage/storage.component.js]:
```
    if (angular.element('#importFile')[0].files[0]) {
      reader.readAsText(angular.element('#importFile')[0].files[0]);
    } else {
      $window.alert("No File Selected!");
    }
```

**Comment by delphiactual** [comment]
Clicking on `Log in to Google Drive` results in the following error
```
TypeError: Cannot read property 'isSignedIn' of null
    at Object.authorize (google-drive-storage.js:100)
    at Object.authorizeGdrive (dimSyncService.factory.js:148)
    at StorageController.vm.driveSync (storage.component.js:65)
    at angular.js:15879
    at callback (angular.js:26994)
    at Scope.$eval (angular.js:18161)
    at Scope.$apply (angular.js:18261)
    at HTMLButtonElement.<anonymous> (angular.js:26999)
    at HTMLButtonElement.dispatch (jquery.js:5206)
    at HTMLButtonElement.elemData.handle (jquery.js:5014)
```

File: src/scripts/storage/storage.component.js (hunk 1 above)
Comment (bhollis) [inline src/scripts/storage/storage.component.js]: Good call

**Comment by bhollis** [comment]
Good catch, I wasn't properly waiting for initialization

**Comment by bhollis** [comment]
screen shot 2017-06-17 at 11 59 33 am

**Comment by bhollis** [comment]
Well, I've hit a bit of a snag - the updated Google Drive APIs I used don't work with Chrome extensions. The previous hack we'd tried (using `chrome.identity`) doesn't work with the new libraries. I could backport everything to the previous API, or I can just disable Google Drive on the extension.

This seems like just another reason to hasten the demise of the chrome extension.

**Comment by bhollis** [comment]
OK, I went a little crazy. I added a feature where the Chrome extension will install a content script for the DIM website. The storage page can communicate with the content script and use it to bring over your data from the extension version (there's an import button).

Note that this will prompt for new permissions on the Chrome extension. With this change, I feel comfortable moving ahead with replacing the Chrome extension with just this content script and a button that opens the website, though.

Alternately, we could assume folks generally don't have much data, and if they want to copy them over they can use the back up / import flow.

screen shot 2017-06-17 at 2 39 58 pm

**Comment by bhollis** [comment]
At this point, I think this is good to go into Beta. @kyleshay @SunburnedGoose, please take a look and validate some of the assumptions and tradeoffs I've made.

**Comment by SunburnedGoose** [comment]
Before we end the extension support, we need to move the app out of beta on the website. Once were versioning the site, then end support for the extension and move forward with the website.

**Comment by bhollis** [comment]
That's right.

**Comment by delphiactual** [comment]
Assuming https://github.com/DestinyItemManager/DIM/pull/1662/commits/20b00dc4fb7c917ab538f465983d682e221273fa caused, when clicking on Vendors...

```
TypeError: Failed to execute 'removeEventListener' on 'EventTarget': 2 arguments required, but only 1 present.
    at storage.component.js:144
    at ChildScope.$broadcast (angular.js:18487)
    at ChildScope.$destroy (angular.js:18102)
    at cleanupLastView (viewDirective.js:90)
    at viewDirective.js:140
    at publicLinkFn (angular.js:8919)
    at lazyCompilation (angular.js:9311)
    at updateView (viewDirective.js:129)
    at Object.configUpdatedCallback [as configUpdated] (viewDirective.js:73)
    at configureUIView (view.js:106)
```

**Comment by delphiactual** [comment]
Also not seeing Google Drive info at all

screenshot 2017-06-17 23 45 39

**Comment by bhollis** [comment]
Whoops, guess I didn't try leaving that page.

As far as Google Drive being missing, that's what I was talking about - the Chrome extension will not support Google Drive, only the website. The website will be able to import data from the extension, though. Soon (once the non-beta website has been released) we will change the extension to be just a link to the website, and there will be no more extension version.

**Comment by delphiactual** [comment]
On import, counts should be updated

**Comment by delphiactual** [comment]
@bhollis this is what I am seeing:

gdrive

**Comment by bhollis** [comment]
Ah, I found it, I wasn't waiting for the `set` promise before refreshing.

**Comment by delphiactual** [comment]
Looks like `'$featureFlags.importExport'` can be removed from `config/webpack.js`

**Comment by adamksib** [comment]
Super excited.  Just a question about cross platform.  User story as follows:

I want to have tags shared on all my devices so that I can do my inventory management during lunch break at work.

Will that be possible assuming I use the web app on both an android and a PC based device?

**Comment by bhollis** [comment]
@adamksib yup, that's the whole point of this!

**Pull request merged**